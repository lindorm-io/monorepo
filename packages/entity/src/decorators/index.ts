export * from "./Column";
export * from "./CreateDateColumn";
export * from "./DeleteDateColumn";
export * from "./Entity";
export * from "./ExpiryDateColumn";
export * from "./Generated";
export * from "./Index_";
export * from "./ManyToMany";
export * from "./ManyToOne";
export * from "./OnCreate";
export * from "./OnDestroy";
export * from "./OneToMany";
export * from "./OneToOne";
export * from "./OnInsert";
export * from "./OnUpdate";
export * from "./OnValidate";
export * from "./PrimaryKey";
export * from "./PrimaryKeyColumn";
export * from "./PrimarySource";
export * from "./Schema";
export * from "./ScopeColumn";
export * from "./UpdateDateColumn";
export * from "./VersionColumn";
export * from "./VersionEndDateColumn";
export * from "./VersionKeyColumn";
export * from "./VersionStartDateColumn";
