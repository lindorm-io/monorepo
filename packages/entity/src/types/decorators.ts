/* eslint-disable @typescript-eslint/no-unsafe-function-type */

import { ZodType } from "zod";
import {
  MetaColumnFallback,
  MetaColumnType,
  MetaGeneratedStrategy,
  MetaIndexOptions,
} from "./metadata";
import { IndexDirection } from "./types";

export type ColumnDecoratorOptions = {
  enum?: any;
  fallback?: MetaColumnFallback;
  max?: number;
  min?: number;
  nullable?: boolean;
  optional?: boolean;
  readonly?: boolean;
  schema?: ZodType;
  type?: MetaColumnType;
};

export type EntityDecoratorOptions = {
  cache?: string;
  database?: string;
  name?: string;
  namespace?: string;
};

export type GeneratedDecoratorOptions = {
  strategy?: MetaGeneratedStrategy;
  length?: number;
  max?: number;
  min?: number;
};

export type HookDecoratorCallback = (entity: any) => void;

export type IndexDecoratorOptions = {
  direction?: IndexDirection;
  name?: string;
  options?: MetaIndexOptions;
  unique?: boolean;
};

export type UniqueDecoratorOptions = {
  name?: string;
};
