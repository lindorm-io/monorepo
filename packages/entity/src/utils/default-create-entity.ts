import { isFunction } from "@lindorm/is";
import { Constructor, DeepPartial } from "@lindorm/types";
import { IEntity } from "../interfaces";
import { globalEntityMetadata } from "./global";
import { parseColumn } from "./private";

export const defaultCreateEntity = <
  E extends IEntity,
  O extends DeepPartial<E> = DeepPartial<E>,
>(
  Entity: Constructor<E>,
  options: O | E = {} as O,
): E => {
  const metadata = globalEntityMetadata.get(Entity);
  const entity = new Entity();

  for (const column of metadata.columns) {
    (entity as any)[column.key] = parseColumn(column, entity, options);

    if ((entity as any)[column.key]) continue;

    (entity as any)[column.key] = isFunction(column.fallback)
      ? column.fallback()
      : column.fallback;
  }

  const hooks = metadata.hooks.filter((h) => h.decorator === "OnCreate");

  for (const hook of hooks) {
    hook.callback(entity);
  }

  return entity;
};
