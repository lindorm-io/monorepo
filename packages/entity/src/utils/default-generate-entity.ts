import { Constructor, DeepPartial } from "@lindorm/types";
import { randomBytes, randomInt, randomUUID } from "crypto";
import { IEntity } from "../interfaces";
import { EntityMetadata, MetaGenerated } from "../types";
import { globalEntityMetadata } from "./global";

const generator = (config: Omit<MetaGenerated, "target">): any => {
  switch (config.strategy) {
    case "date":
      return new Date();

    case "float":
      config.min = config.min ?? 0;
      config.max = config.max ?? 999999;
      return Math.random() * (config.max - config.min) + config.min;

    case "integer":
      config.min = config.min ?? 0;
      config.max = config.max ?? 999999;
      return randomInt(config.min, config.max);

    case "string":
      config.length = config.length ?? 32;
      return randomBytes(config.length).toString("base64url");

    case "uuid":
      return randomUUID();

    default:
      return null;
  }
};

const validator = (
  config: Omit<MetaGenerated, "target">,
  metadata: EntityMetadata,
): void => {
  const column = metadata.columns.find((item) => item.key === config.key);

  if (!column) {
    throw new Error(`Column not found for key: ${config.key}`);
  }

  if (!column.type) return;

  switch (config.strategy) {
    case "date":
      if (column.type === "date") break;
      throw new Error(`Invalid column type for date generation: ${column.type}`);

    case "float":
      if (column.type === "float") break;
      throw new Error(`Invalid column type for int generation: ${column.type}`);

    case "integer":
      if (column.type === "integer") break;
      throw new Error(`Invalid column type for int generation: ${column.type}`);

    case "string":
      if (column.type === "string") break;
      throw new Error(`Invalid column type for string generation: ${column.type}`);

    case "uuid":
      if (column.type === "uuid") break;
      throw new Error(`Invalid column type for uuid generation: ${column.type}`);

    default:
      break;
  }
};

export const defaultGenerateEntity = <E extends IEntity>(
  target: Constructor<E>,
  entity: DeepPartial<E>,
): E => {
  const metadata = globalEntityMetadata.get(target);

  for (const config of metadata.generated) {
    if (entity[config.key]) continue;

    validator(config, metadata);

    (entity as any)[config.key] = generator(config);
  }

  return entity as E;
};
