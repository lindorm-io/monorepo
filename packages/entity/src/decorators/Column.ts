import { isObject, isString } from "@lindorm/is";
import { ColumnDecoratorOptions, MetaColumnType } from "../types";
import { globalEntityMetadata } from "../utils";

export function Column(
  type?: MetaColumnType,
  options?: Omit<ColumnDecoratorOptions, "type">,
): PropertyDecorator;
export function Column(options?: ColumnDecoratorOptions): PropertyDecorator;
export function Column(arg1?: any, arg2?: any): PropertyDecorator {
  return function (target, key) {
    const type = isString<MetaColumnType>(arg1)
      ? arg1
      : isObject<ColumnDecoratorOptions>(arg1)
        ? arg1.type
        : null;

    const options: ColumnDecoratorOptions | undefined = isObject(arg1) ? arg1 : arg2;

    globalEntityMetadata.addColumn({
      target: target.constructor,
      key: key.toString(),
      decorator: "Column",
      enum: options?.enum ?? null,
      fallback: options?.fallback ?? null,
      max: options?.max ?? null,
      min: options?.min ?? null,
      nullable: options?.nullable ?? false,
      optional: options?.optional ?? false,
      readonly: options?.readonly ?? false,
      schema: options?.schema ?? null,
      type: type ?? null,
    });
  };
}
