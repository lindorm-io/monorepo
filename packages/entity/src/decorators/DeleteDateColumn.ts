import { globalEntityMetadata } from "../utils";

export function DeleteDateColumn(): PropertyDecorator {
  return function (target, key) {
    globalEntityMetadata.addColumn({
      target: target.constructor,
      key: key.toString(),
      decorator: "DeleteDateColumn",
      enum: null,
      fallback: null,
      max: null,
      min: null,
      nullable: true,
      optional: false,
      readonly: true,
      schema: null,
      type: "date",
    });
  };
}
