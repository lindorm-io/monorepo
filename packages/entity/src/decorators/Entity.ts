import { EntityDecoratorOptions } from "../types";
import { globalEntityMetadata } from "../utils";

export function Entity(options: EntityDecoratorOptions = {}): ClassDecorator {
  return function (target) {
    globalEntityMetadata.addEntity({
      target,
      decorator: "Entity",
      cache: options?.cache || null,
      database: options?.database || null,
      name: options?.name || null,
      namespace: options?.namespace || null,
    });
  };
}
