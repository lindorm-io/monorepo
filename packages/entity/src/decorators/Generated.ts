import { isObject, isString } from "@lindorm/is";
import { GeneratedDecoratorOptions, MetaGeneratedStrategy } from "../types";
import { globalEntityMetadata } from "../utils";

export function Generated(
  strategy: MetaGeneratedStrategy,
  options?: Omit<GeneratedDecoratorOptions, "strategy">,
): PropertyDecorator;
export function Generated(options?: GeneratedDecoratorOptions): PropertyDecorator;
export function Generated(arg1?: any, arg2?: any): PropertyDecorator {
  return function (target, key) {
    const options = isObject(arg1)
      ? (arg1 as GeneratedDecoratorOptions)
      : ((arg2 ?? {}) as GeneratedDecoratorOptions);
    const strategy = isString(arg1) ? (arg1 as MetaGeneratedStrategy) : options.strategy;
    if (!strategy) {
      throw new Error("Generated strategy is required");
    }
    globalEntityMetadata.addGenerated({
      target: target.constructor,
      key: key as string,
      length: options.length ?? null,
      max: options.max ?? null,
      min: options.min ?? null,
      strategy,
    });
  };
}
