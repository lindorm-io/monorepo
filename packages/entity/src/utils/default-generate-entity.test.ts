import MockDate from "mockdate";
import { Column, Entity, PrimaryKeyColumn } from "../decorators";
import { Generated } from "../decorators/Generated";
import { defaultCreateEntity } from "./default-create-entity";
import { defaultGenerateEntity } from "./default-generate-entity";

const MockedDate = new Date("2024-01-01T08:00:00.000Z");
MockDate.set(MockedDate);

jest.mock("crypto", () => ({
  ...jest.requireActual("crypto"),
  randomBytes: jest
    .fn()
    .mockImplementation(() => Buffer.from("c8ff39528ba451a3a4d52f0726c49524", "hex")),
  randomInt: jest.fn().mockImplementation(() => 123456),
  randomUUID: jest
    .fn()
    .mockImplementation(() => "c8ff3952-8ba4-51a3-a4d5-2f0726c49524")
    .mockImplementationOnce(() => "e8edbe03-f52f-56c6-8d98-82e6961329fe")
    .mockImplementationOnce(() => "0c8c31df-f1cc-5381-97fc-31b3714327de")
    .mockImplementationOnce(() => "2dc92d72-9ce3-50c7-b982-59022ec393f1")
    .mockImplementationOnce(() => "88943033-fef1-5ad8-8cf8-4470be80d197")
    .mockImplementationOnce(() => "1d37a07d-4fd6-5913-85eb-2c7d721cf2a9")
    .mockImplementationOnce(() => "5f2976d2-afda-5e28-9c06-3c04865ea8df"),
}));

describe("defaultGenerateEntity", () => {
  test("should generate standard values", () => {
    @Entity()
    class TestGenerateEntityOptions {
      @PrimaryKeyColumn()
      id!: string;

      @Column("date")
      @Generated("date")
      date!: Date;

      @Column("integer")
      @Generated("integer")
      number!: number;

      @Column("string")
      @Generated("string")
      string!: string;

      @Column("uuid")
      @Generated("uuid")
      uuid!: string;

      @Column("integer")
      @Generated("increment")
      increment!: number;
    }

    expect(
      defaultGenerateEntity(
        TestGenerateEntityOptions,
        defaultCreateEntity(TestGenerateEntityOptions),
      ),
    ).toMatchSnapshot();
  });
});
