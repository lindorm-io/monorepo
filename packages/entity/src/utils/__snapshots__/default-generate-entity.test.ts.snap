// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaultGenerateEntity should generate standard values 1`] = `
TestGenerateEntityOptions {
  "date": 2024-01-01T08:00:00.000Z,
  "id": "e8edbe03-f52f-56c6-8d98-82e6961329fe",
  "increment": null,
  "number": 123456,
  "string": "yP85UoukUaOk1S8HJsSVJA",
  "uuid": "0c8c31df-f1cc-5381-97fc-31b3714327de",
}
`;
